File: intrinsic/skills/internal/skill_init.cc
// Copyright 2023 Intrinsic Innovation LLC

#include "intrinsic/skills/internal/skill_init.h"

#include <memory>
#include <string>
#include <utility>
#include <vector>

#include "absl/algorithm/container.h"
#include "absl/flags/flag.h"
#include "absl/log/log.h"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/string_view.h"
#include "absl/time/clock.h"
#include "absl/time/time.h"
#include "google/protobuf/descriptor.pb.h"
#include "grpc/grpc.h"
#include "grpcpp/channel.h"
#include "grpcpp/security/server_credentials.h"
#include "grpcpp/server.h"
#include "grpcpp/server_builder.h"
#include "intrinsic/icon/release/file_helpers.h"
#include "intrinsic/logging/data_logger_client.h"
#include "intrinsic/motion_planning/proto/motion_planner_service.grpc.pb.h"
#include "intrinsic/skills/internal/skill_registry_client.h"
#include "intrinsic/skills/internal/skill_repository.h"
#include "intrinsic/skills/internal/skill_service_impl.h"
#include "intrinsic/skills/proto/skill_service_config.pb.h"
#include "intrinsic/skills/proto/skills.pb.h"
#include "intrinsic/util/grpc/grpc.h"
#include "intrinsic/util/status/status_macros.h"
#include "intrinsic/world/proto/object_world_service.grpc.pb.h"

namespace intrinsic::skills {
namespace {

using ObjectWorldService = ::intrinsic_proto::world::ObjectWorldService;
using MotionPlannerService =
    ::intrinsic_proto::motion_planning::MotionPlannerService;

absl::StatusOr<std::shared_ptr<
    intrinsic_proto::motion_planning::MotionPlannerService::Stub>>
CreateMotionPlannerServiceStub(absl::string_view motion_planner_service_address,
                               absl::Duration connection_timeout) {
  INTR_ASSIGN_OR_RETURN(const std::shared_ptr<grpc::Channel> channel,
                        CreateClientChannel(motion_planner_service_address,
                                            absl::Now() + connection_timeout));
  return intrinsic_proto::motion_planning::MotionPlannerService::NewStub(
      channel);
}

}  // namespace

absl::Status SkillInit(
    const intrinsic_proto::skills::SkillServiceConfig& service_config,
    absl::string_view data_logger_grpc_service_address,
    absl::string_view world_service_address,
    absl::string_view geometry_service_address,
    absl::string_view motion_planner_service_address,
    absl::string_view skill_registry_service_address,
    int32_t skill_service_port, absl::Duration connection_timeout,
    SkillRepository& skill_repository) {
  // Start DataLogger if the endpoint is configured by flags. Do not fail if
  // the logger is unavailable.
  if (!data_logger_grpc_service_address.empty()) {
    if (auto s = intrinsic::data_logger::StartUpIntrinsicLoggerViaGrpc(
            data_logger_grpc_service_address, connection_timeout);
        !s.ok()) {
      LOG(ERROR) << "Failed to connect to data logger: " << s;
    }
  }

  // Set up world service.
  INTR_ASSIGN_OR_RETURN(
      const std::shared_ptr<grpc::Channel> world_service_channel,
      CreateClientChannel(world_service_address,
                          absl::Now() + connection_timeout));

  std::shared_ptr<ObjectWorldService::StubInterface> object_world_service =
      ObjectWorldService::NewStub(world_service_channel);

  INTR_ASSIGN_OR_RETURN(
      std::shared_ptr<MotionPlannerService::StubInterface>
          motion_planner_service,
      CreateMotionPlannerServiceStub(motion_planner_service_address,
                                     connection_timeout));

  // Set up the skill registry client.
  INTR_ASSIGN_OR_RETURN(
      std::unique_ptr<SkillRegistryClient> skill_registry_client,
      CreateSkillRegistryClient(skill_registry_service_address));

  SkillProjectorServiceImpl project_service(
      skill_repository, object_world_service, motion_planner_service);
  SkillExecutorServiceImpl execute_service(
      skill_repository, object_world_service, motion_planner_service);

  std::string server_address = absl::StrCat("0.0.0.0:", skill_service_port);

  grpc::ServerBuilder builder;
  std::shared_ptr<grpc::ServerCredentials> creds =
      grpc::InsecureServerCredentials();  // NOLINT (insecure)
  builder.AddListeningPort(server_address, creds);
  builder.AddChannelArgument(GRPC_ARG_ALLOW_REUSEPORT, 0);
  builder.RegisterService(&project_service);
  builder.RegisterService(&execute_service);

  if (!service_config.has_skill_description()) {
    LOG(FATAL) << "skill_description was not present in the SkillServiceConfig";
  }
  intrinsic_proto::skills::Skill skill_description =
      service_config.skill_description();

  auto skill_information_service =
      std::make_unique<SkillInformationServiceImpl>(skill_description);
  builder.RegisterService(skill_information_service.get());
  std::unique_ptr<::grpc::Server> server(builder.BuildAndStart());
  if (server == nullptr) {
    LOG(FATAL) << "Cannot create skill service " << server_address;
  }

  std::vector<std::string> skill_names = skill_repository.GetSkillAliases();
  absl::c_sort(skill_names);

  LOG(INFO) << "Available skills: ";
  for (const auto& name : skill_names) {
    LOG(INFO) << "\t" << name;
  }

  LOG(INFO) << "--------------------------------";
  LOG(INFO) << "-- Skill service listening on " << server_address;
  LOG(INFO) << "--------------------------------";

  server->Wait();
  return absl::OkStatus();
}

absl::StatusOr<intrinsic_proto::skills::SkillServiceConfig>
GetSkillServiceConfig(absl::string_view skill_service_config_filename) {
  intrinsic_proto::skills::SkillServiceConfig service_config;
  if (!skill_service_config_filename.empty()) {
    LOG(INFO) << "Reading skill configuration proto from "
              << skill_service_config_filename;
    INTR_ASSIGN_OR_RETURN(
        service_config,
        GetBinaryProto<intrinsic_proto::skills::SkillServiceConfig>(
            skill_service_config_filename));
    LOG(INFO) << "\nUsing skill configuration proto:\n" << service_config;
  }
  return service_config;
}

}  // namespace intrinsic::skills

File: intrinsic/skills/internal/skill_init.py
# Copyright 2023 Intrinsic Innovation LLC
# Intrinsic Proprietary and Confidential
# Provided subject to written agreement between the parties.

"""skill_init provides initialization functions for setting up skill service."""

from concurrent import futures
import time

from absl import logging
from google.protobuf import descriptor_pb2
import grpc
from intrinsic.motion_planning.proto import motion_planner_service_pb2_grpc
from intrinsic.skills.internal import proto_utils
from intrinsic.skills.internal import skill_repository as skill_repo
from intrinsic.skills.internal import skill_service_impl
from intrinsic.skills.proto import skill_service_config_pb2
from intrinsic.skills.proto import skill_service_pb2_grpc
from intrinsic.world.proto import object_world_service_pb2_grpc


def _create_object_world_service_stub(
    address: str, connection_timeout: int
) -> object_world_service_pb2_grpc.ObjectWorldServiceStub:
  channel = grpc.insecure_channel(address)
  # Blocks for the duration of the timeout until the channel is ready.
  grpc.channel_ready_future(channel).result(connection_timeout)
  return object_world_service_pb2_grpc.ObjectWorldServiceStub(channel)


def _create_motion_planner_service_stub(
    address: str, connection_timeout: int
) -> motion_planner_service_pb2_grpc.MotionPlannerServiceStub:
  channel = grpc.insecure_channel(address)
  # Blocks for the duration of the timeout until the channel is ready.
  grpc.channel_ready_future(channel).result(timeout=connection_timeout)
  return motion_planner_service_pb2_grpc.MotionPlannerServiceStub(channel)


def get_skill_service_config(
    filename: str,
) -> skill_service_config_pb2.SkillServiceConfig:
  """Returns the SkillServiceConfig at `skill_service_config_filename`.

  Reads the proto binary file at `skill_service_config_filename` and returns
  the contents. This file must contain a proto binary
  intrinsic_proto.skills.SkillServiceConfig message.

  Args:
    filename: the filename of the binary proto file

  Returns:
    A skill_service_config_pb2.SkillServiceConfig message
  """
  logging.info(
      "Reading skill configuration proto from: %s",
      filename,
  )
  service_config = skill_service_config_pb2.SkillServiceConfig()
  logging.info("\nImporting service_config from file.")
  with open(filename, "rb") as f:
    service_config.ParseFromString(f.read())
  logging.info("\nUsing skill configuration proto:\n%s", service_config)
  return service_config


def skill_init(
    skill_repository: skill_repo.SkillRepository,
    skill_service_config: skill_service_config_pb2.SkillServiceConfig,
    num_threads: int,
    skill_service_port: int,
    world_service_address: str,
    motion_planner_service_address: str,
    geometry_service_address: str,
    connection_timeout: int,
):
  """Starts the skill services on a gRPC server at port `skill_service_port`.

  This server hosts services required for serving a skill including:
    * SkillProjectorServicer
    * SkillExecutorServicer
    * SkillInformationServicer

  Establishes connections to common clients of skills.
  The `connection_timeout` applies to the establishment of each connection, not
  the cumulative connection time.

  The skills services are configured using the proto data contained in the
  service_config.

  If setup passes, this method does not return until the gRPC skill server is
  shutdown. This normally occurs when the process is killed.

  Args:
    skill_repository: The skill repository used to create the skill instance
    skill_service_config: Configuration file for this skill service
    num_threads: The number of thread that the gRPC skill service should use
    skill_service_port: The port that the gRPC service should use
    world_service_address: The address of the world service
    motion_planner_service_address: The address of the motion planner service
    geometry_service_address: The address of the geometry service
    connection_timeout: The connection timeout

  Raises:
    RuntimeError: if skill service fails to use skill_service_port
  """
  server = grpc.server(
      futures.ThreadPoolExecutor(max_workers=num_threads),
      options=(("grpc.so_reuseport", 0),),
  )  # pytype: disable=wrong-keyword-args

  object_world_service = _create_object_world_service_stub(
      world_service_address, connection_timeout
  )
  motion_planner_service = _create_motion_planner_service_stub(
      motion_planner_service_address, connection_timeout
  )

  # Initialize the projector service.
  projector_servicer = skill_service_impl.SkillProjectorServicer(
      skill_repository=skill_repository,
      object_world_service=object_world_service,
      motion_planner_service=motion_planner_service,
  )
  skill_service_pb2_grpc.add_ProjectorServicer_to_server(
      projector_servicer, server
  )

  # Initialize the executor service.
  executor_servicer = skill_service_impl.SkillExecutorServicer(
      skill_repository=skill_repository,
      object_world_service=object_world_service,
      motion_planner_service=motion_planner_service,
  )
  skill_service_pb2_grpc.add_ExecutorServicer_to_server(
      executor_servicer, server
  )

  # Initialize the skill information service if --skill_service_config_filename
  # given (which means we're running a modular skill server).
  if skill_service_config.skill_name or skill_service_config.HasField(
      "skill_description"
  ):
    if (
        skill_service_config.skill_name
        not in skill_repository.get_skill_aliases()
    ):
      raise ValueError(
          "Could not find skill {} in skill modules.".format(
              skill_service_config.skill_name
          )
      )
    logging.info(
        "Adding skill information server with modular skill %s",
        skill_service_config.skill_name,
    )

    parameter_file_descriptor_set = descriptor_pb2.FileDescriptorSet()
    if skill_service_config.parameter_descriptor_filename:
      parameter_file_descriptor_set = descriptor_pb2.FileDescriptorSet()
      with open(skill_service_config.parameter_descriptor_filename, "rb") as f:
        parameter_file_descriptor_set.ParseFromString(f.read())

    if not skill_service_config.HasField("skill_description"):
      skill_object = skill_repository.get_skill(skill_service_config.skill_name)
      skill = proto_utils.proto_from_skill(skill_object)

      proto_utils.add_file_descriptor_set_without_source_code_info(
          skill_object, parameter_file_descriptor_set, skill
      )
    else:
      skill = skill_service_config.skill_description

    skill_info_servicer = skill_service_impl.SkillInformationServicer(skill)
    skill_service_pb2_grpc.add_SkillInformationServicer_to_server(
        skill_info_servicer, server
    )

  # Initialize server with insecure port.
  endpoint = "[::]:{}".format(skill_service_port)
  added_port = server.add_insecure_port(endpoint)
  if added_port != skill_service_port:
    raise RuntimeError(f"Failed to use port {skill_service_port}")
  server.start()

  logging.info("""==========================================================
      """)
  logging.info("--------------------------------")
  logging.info("-- Skill service listening on %s", endpoint)
  logging.info("--------------------------------")

  # Keep server running until an interrupt signal is received.
  try:
    while True:
      # Sleep for a day.
      time.sleep(60 * 60 * 24)
  except KeyboardInterrupt:
    pass
  finally:
    server.stop(None)
